lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -840,8 +840,35 @@ swift::matchWitness(TypeChecker &tc,
     // Try to solve the system disallowing free type variables, because
     // that would resolve in incorrect substitution matching when witness
     // type has free type variables present as well.
-    auto solution = cs->solveSingle(FreeTypeVariableBinding::Disallow);
-    if (!solution)
+    auto solution = cs->solveSingle(FreeTypeVariableBinding::Disallow,
+                                    /* allowFixes */ true);
+
+    // If the types would match but for some other missing conformance, find and
+    // call that out.
+    if (solution && conformance && solution->Fixes.size() == 1 &&
+        solution->Fixes.front()->getKind() == FixKind::AddConformance) {
+      auto fix = (MissingConformance *)solution->Fixes.front();
+      auto type = fix->getNonConformingType();
+      auto protocolType = fix->getProtocol()->getDeclaredType();
+      if (auto memberTy = type->getAs<DependentMemberType>())
+        return RequirementMatch(witness, MatchKind::MissingConformance, type,
+                                protocolType);
+
+      type = type->mapTypeOutOfContext();
+      if (auto typeParamTy = type->getAs<GenericTypeParamType>())
+        if (auto env = conformance->getGenericEnvironment())
+          if (auto assocType = env->mapTypeIntoContext(typeParamTy))
+            return RequirementMatch(witness, MatchKind::MissingConformance,
+                                    assocType, protocolType);
+      if (type->isEqual(conformance->getType())) {
+        if (auto bgt = type->getAs<BoundGenericType>())
+          type = bgt->getDecl()->getDeclaredType();
```

Comment:
What if it’s a non-generic type nested in a generic type? Is this restriction really necessary?